content/03_CreatingObservableSequences.md

```diff
@@ -0,0 +1,1274 @@
+Sub2: 1
+Sub1: 2
+Sub2: 2
+Sub1: 3
+Sub2: 3
+```
+
+We could use this as a way to bridge between some API from which we receive data into the world of Rx. You could imagine writing something of this kind:
+
+```cs
+public class MessageQueueToRx : IDisposable
+{
+    private readonly Subject<string> messages = new();
+
+    public IObservable<string> Messages => messages;
+
+    public void Run()
+    {
+        while (true)
+        {
+            // Receive a message from some hypothetical message queuing service
+            string message = MqLibrary.ReceiveMessage();
+            messages.OnNext(message);
+        }
+    }
+
+    public void Dispose()
+    {
+        message.Dispose();
+    }
+}
+```
+
+It wouldn't be too hard to modify this to use `Observable.Create` instead. But where this approach can become easier is if you need to provide multiple different `IObservable<T>` sources. Imagine we distinguish between different message types based on their content, and publish them through different observables. That's hard to arrange with `Observable.Create` if we still want a single loop pulling messages off the queue.
+
+`Subject<T>` also distributes calls to either `OnCompleted` or `OnError` to all subscribers. Of course, the rules of Rx require that once you have called either of these methods on an `IObserver<T>` (and any `ISubject<T>` is an `IObserver<T>`, so this rule applies to `Subject<T>`) you must not call `OnNext`, `OnError`, or `OnComplete` on that observer ever again. In fact, `Subject<T>` will tolerate calls that break this rule—it just ignores them, so even if your code doesn't quite stick to these rules internally, the `IObservable<T>` you present to the outside world will behave correctly, because Rx enforces this.
+
+`Subject<T>` implements `IDisposable`. Disposing a `Subject<T>` puts it into a state where it will throw an exception if you call any of its methods. The documentation also describes it as unsubscribing all observers, but since a disposed `Subject<T>` isn't capable of producing any further notifications in any case, this doesn't really mean much. (Note that it does _not_ call `OnCompleted` on its observers when you `Dispose` it.) The one practical effect is that its internal field that keeps track of observers is reset to a special sentinel value indicating that it has been disposed, meaning that the one externally observable effect of "unsubscribing" the observers is that if, for some reason, your code held onto a reference to a `Subject<T>` after disposing it, that would no longer keep all the subscribers reachable for GC purposes. If a `Subject<T>` remains reachable indefinitely after it is no longer in use, that in itself is effectively a memory leak, but disposal would at least limit the effects: only the `Subject<T>` itself would remain reachable, and not all of its subscribers.
+
+`Subject<T>` is the most straightforward subject, but there are other, more specialized ones.
+
+## ReplaySubject<T>
```

Comment:
A real-world scenario would be useful

Reply:
Since you've selected text spanning both the `Subject` and `ReplaySubject` sections, it's not clear which bit you're talking about. This is mostly from `Subject`, but since that does in fact have a real example (extracting messages from a queue and delivering them into the world of Rx), I guess you're referring to `ReplaySubject<T>`.

I have never found a use for `ReplaySubject<T>`. There have been times when I thought I had, but later discovered it wasn't a good idea. I don't really know what it's for. (In fact I have a slight suspicion it was ill-conceived.) I'm talking about it for completeness.